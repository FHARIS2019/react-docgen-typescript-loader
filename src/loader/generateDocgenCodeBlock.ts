/*
import * as fs from "fs";
import * as ts from "typescript";

const sourceFile = ts.createSourceFile(
  this.resourcePath,
  fs.readFileSync(this.resourcePath).toString(),
  ts.ScriptTarget.ESNext,
  true / setParentNodes /,
);

// ts.updateSourceFileNode(sourceFile, [...sourceFile.statements,

// ])
const tryBlock = ts.createTry(
  ts.createBlock([], true),
  ts.createCatchClause(
    ts.createVariableDeclaration(
      ts.createIdentifier("__react_docgen_typescript_webpack_plugin_error"),
    ),
    ts.createBlock([]),
  ),
  undefined,
);

const printer = ts.createPrinter({ newLine: ts.NewLineKind.LineFeed });
const result =
  printer.printNode(ts.EmitHint.Unspecified, sourceFile, sourceFile) +
  printer.printNode(ts.EmitHint.Unspecified, tryBlock, sourceFile);

console.log(result);
*/
