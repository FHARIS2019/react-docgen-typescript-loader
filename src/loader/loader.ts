import * as webpack from "webpack";
import { parse as parseComponentDocs } from "react-docgen-typescript/lib/parser.js";
import { createAppError } from "../util";

const loader: webpack.loader.Loader = function(source) {
  // Loaders can support both synchronous and asynchronous modes. Prevent
  // manual execution as a synchronous loader because we will be perform async
  // operations.
  if (!this.async)
    throw createAppError("This loader can not be used synchronously.");

  // Mark the loader as being cacheable since the result should be
  // deterministic.
  this.cacheable(true);

  const componentDocs = parseComponentDocs(this.resourcePath);

  if (componentDocs.length) {
    //
  }

  return source;
};

export default loader;
