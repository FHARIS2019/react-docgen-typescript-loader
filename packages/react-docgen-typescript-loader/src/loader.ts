import webpack from "webpack";
// TODO: Import from "react-docgen-typescript" directly when
// https://github.com/styleguidist/react-docgen-typescript/pull/104 is hopefully
// merged in. Will be considering to make a peer dependency as that point.
import {
  withDefaultConfig,
  withCustomConfig,
  withCompilerOptions,
  ParserOptions,
  FileParser,
  ComponentDoc,
} from "react-docgen-typescript/lib/parser.js";
import LoaderOptions from "./LoaderOptions";
import validateOptions from "./validateOptions";
import generateDocgenCodeBlock from "./generateDocgenCodeBlock";
import { getOptions } from "loader-utils";
import { getProgramProvider, ProgramProvider } from "./language-service";

export default function loader(
  this: webpack.loader.LoaderContext,
  source: string,
) {
  const callback = this.async();
  if (!callback) {
    throw new Error("Expected loader to operate in asynchronous mode.");
  }

  processResource(this, source).then(
    newSource => callback(null, newSource),
    error => callback(error),
  );
}

async function processResource(
  context: webpack.loader.LoaderContext,
  source: string,
): Promise<string> {
  // Mark the loader as being cacheable since the result should be
  // deterministic.
  context.cacheable(true);

  const options: LoaderOptions = getOptions(context) || {};
  validateOptions(options);

  options.docgenCollectionName =
    options.docgenCollectionName || "STORYBOOK_REACT_CLASSES";
  if (typeof options.setDisplayName !== "boolean") {
    options.setDisplayName = true;
  }

  // Convert the loader's flat options into the expected structure for
  // react-docgen-typescript.
  // See: node_modules/react-docgen-typescript/lib/parser.d.ts
  const parserOptions: ParserOptions = {
    propFilter:
      options.skipPropsWithName || options.skipPropsWithoutDoc
        ? {
            skipPropsWithName: options.skipPropsWithName || undefined,
            skipPropsWithoutDoc: options.skipPropsWithoutDoc || undefined,
          }
        : options.propFilter,
  };

  // Configure parser using settings provided to loader.
  // See: node_modules/react-docgen-typescript/lib/parser.d.ts
  let parser: FileParser = withDefaultConfig(parserOptions);
  if (options.tsconfigPath) {
    parser = withCustomConfig(options.tsconfigPath, parserOptions);
  } else if (options.compilerOptions) {
    parser = withCompilerOptions(options.compilerOptions, parserOptions);
  }

  let componentDocs: ComponentDoc[];
  let programProvider: ProgramProvider | undefined;

  if (options.experimentalLanguageServiceProvider) {
    programProvider = getProgramProvider(
      context,
      options,
      source,
      options.experimentalLanguageServiceProvider,
    );
  }

  componentDocs = parser.parse(context.resourcePath, programProvider);

  // Return amended source code if there is docgen information available.
  if (componentDocs.length) {
    return generateDocgenCodeBlock({
      filename: context.resourcePath,
      source,
      componentDocs,
      docgenCollectionName: options.docgenCollectionName,
      setDisplayName: options.setDisplayName,
    });
  }

  // Return unchanged source code if no docgen information was available.
  return source;
}
