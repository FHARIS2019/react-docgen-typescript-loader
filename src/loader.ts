import * as webpack from "webpack";
import {
  withDefaultConfig,
  withCustomConfig,
  withCompilerOptions,
  ParserOptions,
  FileParser,
} from "react-docgen-typescript/lib/parser.js";
import LoaderOptions from "./LoaderOptions";
import validateOptions from "./validateOptions";
import generateDocgenCodeBlock from "./generateDocgenCodeBlock";

const loader: webpack.loader.Loader = function(source) {
  const callback = this.async();
  try {
    // Mark the loader as being cacheable since the result should be
    // deterministic.
    this.cacheable(true);

    const options: LoaderOptions = this.query || {};
    validateOptions(options);
    options.docgenCollectionName =
      options.docgenCollectionName || "STORYBOOK_REACT_CLASSES";

    // Convert loader's flat options into expected structure for
    // react-docgen-typescript.
    // See: node_modules/react-docgen-typescript/lib/parser.d.ts
    const parserOptions: ParserOptions = {
      propFilter:
        options.skipPropsWithName || options.skipPropsWithoutDoc
          ? {
              skipPropsWithName: options.skipPropsWithName || undefined,
              skipPropsWithoutDoc: options.skipPropsWithoutDoc || undefined,
            }
          : undefined,
    };

    const newSource = (() => {
      // Configure parser using settings provided to loader.
      // See: node_modules/react-docgen-typescript/lib/parser.d.ts
      let parser: FileParser = withDefaultConfig(parserOptions);
      if (options.tsconfigPath)
        parser = withCustomConfig(options.tsconfigPath, parserOptions);
      else if (options.compilerOptions)
        parser = withCompilerOptions(options.compilerOptions, parserOptions);

      const componentDocs = parser.parse(this.resourcePath);

      // Return amended source code if there is docgen information available.
      if (componentDocs.length) {
        return generateDocgenCodeBlock(
          this.resourcePath,
          source as string,
          componentDocs,
          options.docgenCollectionName,
        );
      }

      // Return unchanged source code if no docgen information was available.
      return source;
    })();

    if (!callback) return newSource;
    callback(null, newSource);
    return;
  } catch (e) {
    if (callback) {
      callback(e);
      return;
    }
    throw e;
  }
};

export default loader;
