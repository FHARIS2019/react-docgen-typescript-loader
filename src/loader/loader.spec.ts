import * as path from "path";
import * as webpack from "webpack";
import loader from "./loader";

const mockLoaderContextAsyncCallback = jest.fn();
const mockLoaderContextCacheable = jest.fn();
const mockLoaderContextResourcePath = jest.fn();

beforeEach(() => {
  mockLoaderContextAsyncCallback.mockReset();
  mockLoaderContextCacheable.mockReset();
  mockLoaderContextResourcePath.mockReset();
  mockLoaderContextResourcePath.mockImplementation(() =>
    path.resolve(__dirname, "./__fixtures__/SimpleComponent.tsx"),
  );
});

it("throws error if used synchronously", () => {
  // Loader expects to have "this" bound to an instance of LoaderContext.
  const synchronouslyLoaderContext = {} as webpack.loader.LoaderContext & {
    loader: webpack.loader.Loader;
  };
  synchronouslyLoaderContext.loader = loader;

  expect(() =>
    synchronouslyLoaderContext.loader("", ""),
  ).toThrowErrorMatchingSnapshot();
});

it("marks the loader as being cacheable", () => {
  executeLoaderWithBoundContext();

  expect(mockLoaderContextCacheable.mock.calls[0][0]).toEqual(true);
});

// Execute loader with its "this" set to an instance of LoaderContext.
function executeLoaderWithBoundContext() {
  const mockLoaderContext = {} as webpack.loader.LoaderContext & {
    loader: webpack.loader.Loader;
  };

  mockLoaderContext.async = mockLoaderContextAsyncCallback;
  mockLoaderContext.cacheable = mockLoaderContextCacheable;
  mockLoaderContext.loader = loader;
  mockLoaderContext.resourcePath = mockLoaderContextResourcePath();

  mockLoaderContext.loader("// Original Source Code", "");
}
