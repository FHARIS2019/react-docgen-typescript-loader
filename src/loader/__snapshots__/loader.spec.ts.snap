// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throws error if used synchronously 1`] = `"react-docgen-typescript-webpack-plugin: This loader can not be used synchronously."`;
